proxy/coco/src/project/checkout.rs

```diff
@@ -66,78 +82,137 @@ pub enum Ownership {
-        builder: &mut git2::build::RepoBuilder,
-    ) -> Result<git2::Repository, git2::Error> {
-        builder.remote_create(|repo, _remote_name, url| repo.remote(config::RAD_REMOTE, url));
-        git2::build::RepoBuilder::clone(builder, &url.to_string(), path)
+    ) -> Result<(git2::Repository, Remote<LocalUrl>), Error>
+    where
+        F: CanOpenStorage + 'static,
+    {
+        let rad = Remote::rad_remote(
+            url,
+            Refspec {
+                src: refspec_pattern!("refs/heads/*"),
+                dst: refspec_pattern!("refs/remotes/rad/*"),
+                force: Force::True,
+            },
+        );
+        clone(path, open_storage, rad)
     }
 
     /// See [`Checkout::run`].
-    fn remote(
+    fn remote<F>(
+        open_storage: F,
         handle: &str,
         peer: PeerId,
         url: LocalUrl,
         default_branch: &OneLevel,
         path: &path::Path,
-        builder: &mut git2::build::RepoBuilder,
-    ) -> Result<git2::Repository, git2::Error> {
+    ) -> Result<(git2::Repository, Remote<LocalUrl>), Error>
+    where
+        F: CanOpenStorage + Clone + 'static,
+    {
         let name =
             RefLike::try_from(format!("{}@{}", handle, peer)).expect("failed to parse remote name");
-        {
-            builder.remote_create(move |repo, _remote_name, url| {
-                let mut remote = Remote::new(url, name.as_str().to_string());
-                let heads: FlatRef<PeerId, _> = FlatRef::heads(PhantomData, peer).with_name(
-                    RefspecPattern::try_from("heads/*").expect("'heads/*' failed to parse"),
-                );
-                let remotes: FlatRef<RefLike, _> = FlatRef::heads(PhantomData, name.clone());
-                remote.fetch_spec = Some(remotes.refspec(heads, Force::True).boxed());
-                remote.create(repo)
-            });
-        }
 
-        let repo = git2::build::RepoBuilder::clone(builder, &url.to_string(), path)?;
+        let remote = Remote::new(url.clone(), name.clone()).with_fetchspecs(vec![Refspec {
+            src: Reference::heads(Flat, peer),
+            dst: GenericRef::heads(Flat, name),
+            force: Force::True,
+        }]);
+
+        let (repo, _) = clone(path, open_storage.clone(), remote)?;
 
         // Create a rad remote and push the default branch so we can set it as the
         // upstream.
-        {
-            let mut remote = Remote::rad_remote(url, None).create(&repo)?;
-            remote.push(&[&format!("refs/heads/{}", default_branch.as_str())], None)?;
-        }
+        let rad = {
+            // Create a fetchspec `refs/heads/*:refs/remotes/rad/*`
+            let fetchspec = Refspec {
+                src: GenericRef::<_, RefLike, _>::heads(Flat, None),
+                dst: refspec_pattern!("refs/remotes/rad/*"),
+                force: Force::True,
+            };
+            let mut rad = Remote::rad_remote(url, fetchspec);
+            rad.save(&repo)?;
```

Comment:
Is it necessary to add the source as a remote in the git config? Isn’t this what the git includes are for?

Reply:
It's necessary for the checkout to perform correctly. It could be that we could use the `include` here though.